marlowe-playground-client/src/Halogen/BlocklyCommons.purs

```diff
@@ -44,31 +54,128 @@ blocklyEvents toAction workspace =
-    -- The move event only changes the unsaved status if the parent has changed (either by attaching or detaching
-    -- one block into another)
-    (BT.Move ev) -> when (useEvents && newParentId ev /= oldParentId ev) $ raise codeChange
-    (BT.FinishLoading _) -> pure unit
-    -- The create event by itself does not modify the contract. It is modified once it's attached or detached
-    -- from a parent, and that is covered by the Move event
-    (BT.Create _) -> pure unit
+  HalogenM { eventsWhileDragging :: Maybe (List BlocklyEvent) | state } action slots message m Unit
+detectCodeChanges codeChange event = do
+  mDraggingEvents <- use _eventsWhileDragging
+  case mDraggingEvents of
+    Nothing ->
+      -- If we are not inside a drag and drop, let's evaluate the event directly
+      if (doesEventModifiesContract event) then
+        H.raise codeChange
+      else
+        -- If the event starts a drag and drop, store it in the state
+        when (isDragStart event) $ assign _eventsWhileDragging (Just Nil)
+    Just eventsWhileDragging ->
+      -- If we are inside a drag and drop, accumulate the events and analyze them all together
+      if (not $ isDragStop event) then
+        assign _eventsWhileDragging (Just (event : eventsWhileDragging))
+      else do
+        let
+          hasChanged = foldl (\accu ev -> accu || doesEventModifiesContract ev) false eventsWhileDragging
+        when hasChanged $ H.raise codeChange
+        assign _eventsWhileDragging Nothing
+  where
+  doesEventModifiesContract :: BlocklyEvent -> Boolean
+  doesEventModifiesContract = case _ of
+    (BT.Change _) -> true
+    (BT.Move ev) -> newParentId ev /= oldParentId ev
+    _ -> false
+
+-- | This function listens to blockly events and debounces the event listener so that the resulting
+-- | Aff is completed after `time` milliseconds without events.
+-- | NOTE: The part of this code that depends on blockly is rather minimal (addChangeListener workspace and
+-- |       removeChangeListener workspace). We don't care about the contents of the event, just the fact that
+-- |       it is fired, so we can debounce it.
+-- |       If we need a debounce functionality later on, see if it makes sense to refactor this. I didn't do it
+-- |       for the PR that introduced this code as it's not clear at the moment wether the debounce should be
+-- |       tied to event listeners or if we can do it with plain Aff's
+waitForEvents :: Workspace -> Int -> Aff Unit
+waitForEvents workspace time = liftEffect (Ref.new Nothing) >>= waitForEventsAndCleanResources
+  where
+  -- Make sure that the event listener is removed no matter what
+  waitForEventsAndCleanResources :: Ref (Maybe EventListener) -> Aff Unit
+  waitForEventsAndCleanResources listenerRef =
+    finally
+      (liftEffect $ removeListener listenerRef)
+      (waitForEvent listenerRef)
+
+  waitForEvent :: Ref (Maybe EventListener) -> Aff Unit
+  waitForEvent listenerRef = makeAff resolver
```

Comment:
maybe you could pull this out as a utility function `debouce` somewhere?

Reply:
As I wrote in the comment, I think I'd like to refactor this to be a general utility function once we have a second use for debounce, as it's not currently clear if we want to have this for plain Aff (or how to do that), or with Event listeners